packages/flutter_tools/lib/src/build_system/targets/web.dart

```diff
@@ -732,3 +735,281 @@ function onlineFirst(event) {
+      for (final String noMd5Path in imagesMap.keys) {
+        final String md5Path = imagesMap[noMd5Path] ?? '';
+        if (mainJsFileContent.contains(noMd5Path)) {
+          final String newMainFileContent = mainJsFileContent.replaceAll(noMd5Path, md5Path);
+          mainJsFile.writeAsStringSync(newMainFileContent);
+          mainJsFileContent = newMainFileContent;
+        }
+        /// replace images new path in xxx.part.js
+        for (final String jsPath in jsPartArr) {
+          // final File jsFile = globals.fs.file(jsPath);
+          final File? jsFile = _convertFullPathToFile(environment, jsPath);
+          if (jsFile != null) {
+            final String jsFileContent = jsFile.readAsStringSync();
+            if (jsFileContent.contains(noMd5Path)) {
+              final String newFileContent = jsFileContent.replaceAll(noMd5Path, md5Path);
+              jsFile.writeAsStringSync(newFileContent);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /// use js path to add sourcemap hash
+  void _sourceMapAddHash(Environment environment, String noMd5JsPath, String md5JsPath, String hash) {
+    final String sourceMapPath = '$noMd5JsPath.map';
+    final File? sourceMapFile = _convertFullPathToFile(environment, sourceMapPath);
+
+    final String jsName = _getFileNameFromPath(noMd5JsPath);
+    final String jsNameWithMd5 = _getFileNameFromPath(md5JsPath);
+    final String sourceMapName = '$jsName.map';
+    final String sourceMapNameWithHash = '$jsNameWithMd5.map';
+
+    final File? jsFile = _convertFullPathToFile(environment, noMd5JsPath);
+    if (jsFile != null) {
+      final String jsContent = jsFile.readAsStringSync();
+      final String newJsContent = jsContent.replaceAll(sourceMapName, sourceMapNameWithHash);
+      jsFile.writeAsStringSync(newJsContent);
+    }
+    // sourceMapFile.rename('$md5JsPath.map');
+    if (sourceMapFile != null && sourceMapFile.existsSync()) {
+      sourceMapFile.copy('$md5JsPath.map');
+    }
+  }
+
+  /// xxx.part.js add hash, for example main.dart.js_1.part.js --> main.dart.js_1.part.[hash].js
+  Future<String> jsAddHash(Environment environment) async{
+    final List<String> list = resourcesMap['js'] ?? <String>[];
+    final List<String> partJsPathArr = <String>[];
+    String mainJsPath = '';
+    String mainJsContent = '';
+    final RegExp partReg = RegExp(r'main\.(.+)\.part\.js$');
+    final RegExp mainReg = RegExp(r'main.dart.js$');
+    String newMainJsPath = '';
+    String mainHash = '';
+    /// filter main.dart.js and xxx.part.js
+    for (final String path in list) {
+      if (partReg.hasMatch(path)) {
+        partJsPathArr.add(path);
+      } else if (mainReg.hasMatch(path)) {
+        mainJsPath = path;
+      }
+    }
+    // final File mainJsFile = globals.fs.file(mainJsPath);
+    final File? mainJsFile = _convertFullPathToFile(environment, mainJsPath);
+    if (mainJsFile != null) {
+      mainHash = await _getFileMd5(mainJsFile);
```

Comment:
This hashes the main file prior to replacing its content in `_dealPartJs`, which means that this is not its true hash. Not sure how problematic this is, but it something to be aware of.